dali/python/nvidia/dali/pipeline.py

```diff
@@ -990,3 +996,113 @@ def iter_setup(self):
         For example, one can use this function to feed the input
         data from NumPy arrays."""
         pass
+
+
+def _discriminate_args(func, **func_kwargs):
+    """Split args on those applicable to Pipeline constructor and the decorated function."""
+    func_argspec = inspect.getfullargspec(func)
+    ctor_argspec = inspect.getfullargspec(Pipeline.__init__)
+
+    ctor_args = {}
+    fn_args = {}
+
+    for farg in func_kwargs.items():
+        is_ctor_arg = farg[0] in ctor_argspec.args or farg[0] in ctor_argspec.kwonlyargs
+        is_fn_arg = farg[0] in func_argspec.args or farg[0] in func_argspec.kwonlyargs
+        if is_fn_arg:
+            fn_args[farg[0]] = farg[1]
+            if is_ctor_arg:
+                print(
+                    "Warning: the argument `{}` shadows a Pipeline constructor argument of the same name.".format(
+                        farg[0]))
+        elif is_ctor_arg:
+            ctor_args[farg[0]] = farg[1]
+        else:
+            fn_args[farg[0]] = farg[1]
+
+    for farg in fn_args.items():
+        if farg[0] not in func_argspec.args and farg[0] not in func_argspec.kwonlyargs:
+            raise TypeError(
+                "Using non-explicitly declared arguments in graph-defining function is not allowed. "
+                "Please remove `{}` argument or declare it explicitly in the function signature.".format(
+                    farg[0]))
```

Comment:
```suggestion
            raise ValueError("`{}` is neither an argument of the graph definition function nor of `Pipeline.__init__`".format(farg[0]))
```

Reply:
We can explicitly check **kwargs (see above), but we can still produce a nice error here - but with a slightly different descirption.